### ISSUE TITLE
Keras HDF5 with int8/bfloat16 support

### ISSUE THREAD

**Issue by tonyreina** [tensorflow/tensorflow#26759]
**System information**
- TensorFlow version (you are using): TensorFlow 1.12
- Are you willing to contribute it (Yes/No): Yes

**Describe the feature and the current behavior/state.**

The TensorFlow protobuf saved model format is aware of reduced precision (int8), but I don't think the Keras HDF5 saved format is.  

**Will this change the current api? How?**

I'd like to have a way to convert a Keras HDF5 model to int8. So when I do a `tf.keras.model.predict()` it will use the INT8 operations rather than the FP32.

**Who will benefit with this feature?**

Anyone using the tf.Keras API.

**Any Other info.**
For example, TF Lite can do this:  https://www.tensorflow.org/lite/performance/post_training_quantization

**Comment by tonyreina** [comment]
Just circling back on this.  Is there a way to use INT8 or BFLOAT16 with the Keras HDF5 format? If not, will this be implemented? Thanks.

**Comment by jvishnuvardhan** [comment]
@tonyreina Could you describe a use case for this feature? Thanks!

**Comment by tonyreina** [comment]
Sure.  

So hardware is moving toward using BFLOAT16 and INT8 operations. For example, Intel Xeon supports DL Boost which is an INT8 precision.  We have pre-trained Keras models in FP32 which we've converted to TensorFlow protobuf models to take advantage of INT8, but that's going to require users to convert every model from a Keras HDF5 saved file to TensorFlow protobuf. And, it means we can't use model.predict() with the tf. Keras API. So it makes it hard to justify the use of Keras for these use cases if the Keras saved format needs to be changed to protobufs.

I believe Google already has BFLOAT16 support for TPUs (https://cloud.google.com/tpu/docs/bfloat16) but-- again-- I'm not aware of any Keras support for this. Perhaps it exists or is in development but I'm not seeing how to use it.

If TensorFlow 2.0 is moving to use Keras as the preferred API, it would help to add support for these lower precision formats.

**Comment by karmel** [comment]
We would recommend you use the TF SavedModel format in this case. In TF 2.0, this will be the default saving format; in the meantime, you can use tf.keras.experimental.export_saved_model() or model.save_weights(...save_format='tf')

**Comment by tonyreina** [comment]
I see. What will happen with tf.keras.model.predict?  Will I be able to load a TF SavedModel via it and do inference that way?

**Comment by karmel** [comment]
Yes, you should be able to use tf.keras.experimental.load_from_saved_model to reload the saved model back into python.

**Comment by tonyreina** [comment]
Thanks Karmel.  That's good to hear.  Is the idea that eventually the tf.keras.models.save_model will actually use the SavedModel protobuf format instead of the Keras HDF5 format?  I'm thinking that the typical user will probably use the save_model and load_model functions rather than the experimental ones.

**Comment by karmel** [comment]
Yes, we anticipate making the TF format the default for tf.keras.Model.save in TF 2.0.

**Comment by jvishnuvardhan** [comment]
@tonyreina Is this still an issue? Can you check with `TF2.0` and let us know whether the issue persists with latest TF version. Thanks!

**Comment by tonyreina** [comment]
It should not be.  I believe that with TF 2.0 the Keras API is now saved as a protobuf. So that should handle the reduced precision formats.

**Comment by jvishnuvardhan** [comment]
I am closing the issue as it was resolved. Please feel free to open it if the issue persists again. Thanks!
